test/aleph/classpath_test.clj

```diff
@@ -0,0 +1,42 @@
+(ns aleph.classpath-test
+  (:require  [clojure.test      :refer [deftest testing is]]
+             [aleph.http        :as http]
+             [manifold.deferred :as d]
+             [signal.handler    :refer [with-handler]])
+  (:import [io.netty.util.concurrent Future]
+           [java.lang.management ManagementFactory]
+           [java.util.concurrent CompletableFuture]))
+
+(defn- operation-complete
+  "Stubs for `GenericFutureListener/operationComplete` which
+  returns a completed `CompletableFuture` containing either
+  `true` or `false` in case of Throwable."
+  [^CompletableFuture result ^Future f d]
+  (try (d/success! d (.getNow f))
+       (.complete result true)
+       d
+       (catch Throwable _ (do (.complete result false) d))))
+
+(defn pid
+  "Gets this process' PID."
+  []
+  (let [pid (.getName (ManagementFactory/getRuntimeMXBean))]
+    (->> pid
+         (re-seq #"[0-9]+")
+         (first)
+         (Integer/parseInt))))
+
+(deftest test-classpath
+  (testing "classpath: ensure the class loader is always a DynamicClassLoader"
+    (let [result (CompletableFuture.)]
+      (intern 'aleph.netty.impl 'operation-complete (partial operation-complete result))
+      (require 'aleph.netty :reload)
+      (let [server (http/start-server
+                    (constantly {:body "ok"})
+                    {:port 9999})]
+        (with-handler :term
+          (.close ^java.io.Closeable server))
+        (.exec (Runtime/getRuntime) (format "kill -SIGTERM %s" (pid)))
+        (is @result)
+        (require 'aleph.netty.impl :reload)
+        (require 'aleph.netty :reload)))))
```

Comment:
This interning and re-requiring seems more complicated than it needs to be. Since we know `aleph.netty.impl/operation-complete` exists, we can use `alter-var-root` or `with-redefs` to temporarily change its root binding, then set it back when done. It's effectively equivalent, but with no need to reload anything. I'm not seeing any benefit to the macro/intern/reload setup over `with-redefs`, but please let me know if I'm missing something. (`with-redefs` is not multithread-safe, but neither is this, since concurrent tests would use the interned version.)

(Incidentally, I have a blog post coming out about in the next week or two highlighting a much better technique for exactly this sort of test redefinition, I'm just waiting on someone to get back to me.)

As I mentioned elsewhere, let's switch to the SIGINT over SIGTERM to be safer.

Reply:
I will work on simplifying this test.

Reply:
> I'm not seeing any benefit to the macro/intern/reload setup over with-redefs, but please let me know if I'm missing something. (with-redefs is not multithread-safe, but neither is this, since concurrent tests would use the interned version.)

As long as we are not using a macro for `operation-complete` we can use `with-redefs`.